src/main/java/org/apache/ibatis/migration/PropertiesEnvironmentLoader.java

```diff
@@ -0,0 +1,131 @@
+/**
+ *    Copyright 2010-2018 the original author or authors.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package org.apache.ibatis.migration;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Properties;
+import java.util.Set;
+
+import org.apache.ibatis.migration.options.SelectedPaths;
+
+public class PropertiesEnvironmentLoader implements EnvironmentLoader {
+
+  protected enum SETTING_KEY {
+    time_zone,
+    delimiter,
+    script_char_set,
+    full_line_delimiter,
+    send_full_script,
+    auto_commit,
+    remove_crs,
+    ignore_warnings,
+    driver_path,
+    driver,
+    url,
+    username,
+    password,
+    hook_before_up,
+    hook_before_each_up,
+    hook_after_each_up,
+    hook_after_up,
+    hook_before_down,
+    hook_before_each_down,
+    hook_after_each_down,
+    hook_after_down,
+    hook_before_new,
+    hook_after_new
+  }
+
+  protected static final List<String> SETTING_KEYS;
```

Comment:
I thought that users might create a custom loader by subclassing this built-in one.

Reply:
Ah that makes sense but I think if that's the case we should move this to something common.
This is the keys though, it is an implementation detail that we have changed before so depending on it might not be a good idea